### PR TITLE
fix: Investigate Gemma3 1B decoder output discrepancy

### PR DESCRIPTION
**Pull request by brb-nv**
## Description

This MR investigates discrepancy noted by stakeholders in Gemma3 decoder outputs when compared to reference HF implementation.

Key findings:
- Gemma3's sliding window isn't left-inclusive in their HF reference implementation. TRTLLM's attention backends (trtllm, flashinfer, vanilla attention(with changes in this MR)) are left-inclusive.
- Subtle difference in position embedding math introduces non-trivial delta in logits. If applying position embeddings is turned off, atol and rtol could be brought down significantly.

Changes in this MR:
- Unit test with a miniature version of Gemma3 1B. Has one local layer (window size=4) and one global layer.
- Improved functionality in vanilla backend: q_scaling, sliding_window. Using vanilla backend has been very useful for debugging.
- Disabling KV cache reuse when working with Gemma3 because the new sliding window size 511 passed on to the attention layer isn't playing well with fmhaRunner.
```
$ python3 examples/pytorch/quickstart_advanced.py --model_dir /home/scratch.trt_llm_data/llm-models/gemma/gemma-3-1b-it/ --max_tokens 2048 --max_seq_len 2048 --prompt "Write me a really long Shakespeare play." --disable_kv_cache_reuse
```

## Test Coverage

```
$ pytest tests/unittest/_torch/modeling/test_modeling_gemma3.py::TestGemma3::test_gemma3_allclose_to_hf[backend:trtllm] -s -v
$ pytest tests/unittest/_torch/modeling/test_modeling_gemma3.py::TestGemma3::test_gemma3_allclose_to_hf[backend:vanilla] -s -v
$ pytest tests/unittest/_torch/modeling/test_modeling_gemma3.py::TestGemma3::test_gemma3_sanity -s -v
$ pytest tests/integration/defs/accuracy/test_llm_api_pytorch.py::TestGemma3_1BInstruct::test_auto_dtype -s -v
```

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_gemma3.py [hunk 1]
```diff
@@ -107,7 +107,8 @@ def forward(
         **kwargs,
     ) -> torch.Tensor:
 
-        attention_window_size = self.attention_window_size or attn_metadata.max_seq_len
+        attention_window_size = self.attention_window_size or getattr(
+            attn_metadata, "max_seq_len", None)
```
Comment (jaedeok-nvidia) [inline tensorrt_llm/_torch/models/modeling_gemma3.py:111]: Is there any specific reason to set `attention_window_size` to a non-None value? Otherwise, can we remove this and just let `Attention.forward` handle it? [TrtllmAttention](https://github.com/NVIDIA/TensorRT-LLM/blob/04fa6c0cfc52303ec692a0cb70657e51d0db4781/tensorrt_llm/_torch/attention_backend/trtllm.py#L223) already handles it internally. I am not sure how the other backends handle this though. If this is common logic, then handling the None value inside `Attention.forward` would be more maintainable.

File: tests/unittest/_torch/modeling/test_modeling_gemma3.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    "num_hidden_layers": 2,  # Modified for testing.
+    "num_key_value_heads": 1,
+    "pad_token_id": 0,
+    "query_pre_attn_scalar": 256,
+    "rms_norm_eps": 1e-06,
+    "rope_local_base_freq": 10000,
+    "rope_scaling": None,
+    "rope_theta": 1000000,
+    "sliding_window": 4,  # Modified for testing.
+    "sliding_window_pattern": 2,  # Modified for testing.
```
[the remark this answers is not in the document]
Comment (brb-nv) [reply tests/unittest/_torch/modeling/test_modeling_gemma3.py:48]: It determines how often a `global` layer appears. In this test, `layer_1` (zero-indexed) is global layer.

**Review by amukkara — APPROVED** [review]

**Review by byshiue — APPROVED** [review]
